03_Modules/OcppRouter/src/module/webhook.dispatcher.ts

```diff
@@ -0,0 +1,288 @@
+    this._onCloseCallbacks.has(identifier)
+      ? this._onCloseCallbacks.get(identifier)!.push(onCloseCallback)
+      : this._onCloseCallbacks.set(identifier, [onCloseCallback]);
+  }
+
+  addOnMessageCallback(
+    identifier: string,
+    onMessageCallback: OnMessageCallback,
+  ): void {
+    this._onMessageCallbacks.has(identifier)
+      ? this._onMessageCallbacks.get(identifier)!.push(onMessageCallback)
+      : this._onMessageCallbacks.set(identifier, [onMessageCallback]);
+  }
+
+  addSentMessageCallback(
+    identifier: string,
+    sentMessageCallback: OnSentMessageCallback,
+  ): void {
+    this._sentMessageCallbacks.has(identifier)
+      ? this._sentMessageCallbacks.get(identifier)!.push(sentMessageCallback)
+      : this._sentMessageCallbacks.set(identifier, [sentMessageCallback]);
+  }
+
+  async register(identifier: string) {
+    this._loadSubscriptionsForConnection(identifier).then(() => {
+      this._onConnectionCallbacks.get(identifier)?.forEach(async (callback) => {
+        callback();
+      });
+    });
+    this._identifiers.add(identifier);
+  }
+
+  async deregister(identifier: string) {
+    this._onCloseCallbacks.get(identifier)?.forEach((callback) => {
+      callback();
+    });
+    this._identifiers.delete(identifier);
+  }
+
+  async dispatchMessageReceived(
+    identifier: string,
+    message: string,
+    info?: Map<string, string>,
+  ) {
+    this._onMessageCallbacks.get(identifier)?.forEach((callback) => {
+      callback(message, info);
+    });
+  }
+
+  async dispatchMessageSent(
+    identifier: string,
+    message: string,
+    error?: any,
+    info?: Map<string, string>,
+  ) {
+    this._sentMessageCallbacks.get(identifier)?.forEach((callback) => {
+      callback(message, error, info);
+    });
+  }
+
+  private async _refreshSubscriptions() {
+    if (this._identifiers.size === 0) {
+      return;
+    }
+    this._logger.debug(
+      `Refreshing subscriptions for ${this._identifiers.size} identifiers`,
+    );
+    this._identifiers.forEach((identifier) =>
+      this._loadSubscriptionsForConnection(identifier),
+    );
+  }
+
+  /**
+   * Loads all subscriptions for a given connection into memory
+   *
+   * @param {string} connectionIdentifier - the identifier of the connection
+   * @return {Promise<void>} a promise that resolves once all subscriptions are loaded
+   */
+  private async _loadSubscriptionsForConnection(connectionIdentifier: string) {
+    const onConnectionCallbacks: OnConnectionCallback[] = [];
+    const onCloseCallbacks: OnCloseCallback[] = [];
+    const onMessageCallbacks: OnMessageCallback[] = [];
+    const sentMessageCallbacks: OnSentMessageCallback[] = [];
+
+    const subscriptions =
+      await this._subscriptionRepository.readAllByStationId(
+        connectionIdentifier,
+      );
+
+    for (const subscription of subscriptions) {
+      if (subscription.onConnect) {
+        onConnectionCallbacks.push(this._onConnectionCallback(subscription));
```

Comment:
I wonder if this makes sense, shouldn't all instances hold all subscriptions for `onConnection`? Since onConnect fires when the charger connects, shouldn't all instances hold all onConnection subscriptions, else you only pick them up when the charger is already connected, since you filter by the connection that the instance is connected to.

Reply:
Since the subscriptions get loaded when the charger connects, and before the subscription actions are triggered, this should be fine. I misinterpreted when the subscriptions get loaded.